### ISSUE TITLE
Remove arguments passed by decorators from function type hint

### ISSUE THREAD

**Issue by qumusabel** [microsoft/pyright#1758]
**Describe the bug**
I can't seem to find a way to remove the arguments that should be passed by the decorator from the function type hint.

**To Reproduce**
This code should reproduce the issue:
```py
from functools import wraps
import rich

def get_console(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        console = rich.get_console()
        return func(console, *args, **kwargs)
    return wrapper

@get_console # pass the global console object to any function that needs it
def error(console: Console, msg: str):
    console.print(r"[red][bold]Error![/bold] {}[/]".format(msg))
```
Now when I type `error` the hint is `error(console: Console, msg: str) -> None`, and `error("Something went wrong")` is highlighted with `Argument missing for parameter "msg"` (and type mismatch).

**Expected behavior**
The argument which is passed to the function by the decorator should be hidden from the type hint. In the example, I want `error(msg: str) -> None` to be the type hint.

**VS Code extension or command-line**
I am using coc-pyright for coc.nvim.

P.S. I'm sorry if this is a duplicate or unrelated to pyright.

**Comment by erictraut** [comment]
Until recently, there was no way to provide type annotations to handle this pattern. The good news is that [PEP 612](https://www.python.org/dev/peps/pep-0612/), which was recently accepted, was created specifically for this use case.

Here's how you would use it:

```python
from typing import Callable, TypeVar
from typing_extensions import Concatenate, ParamSpec

_P = ParamSpec("_P")
_T = TypeVar("_T")

def get_console(func: Callable[Concatenate[Console, _P], _T]) -> Callable[_P, _T]:
    @wraps(func)
    def wrapper(*args: _P.args, **kwargs: _P.kwargs):
        console = rich.get_console()
        return func(console, *args, **kwargs)
    return wrapper
```

Pyright has full support for PEP 612, so you can use this today. You'll need to import `Concatenate` and `ParamSpec` from `typing_extensions` unless you're using alpha versions of Python 3.10 (in which case you can import these symbols directly from `typing`).

**Comment by qumusabel** [comment]
Works like a charm in the editor, thank you very much. However, it doesn't seem to work when the code is actually executed. It throws an ImportError from typing_extensions (but in the editor it shows them in autocomplete). And the PyPI packages for both typing and typing_extensions haven't been updated for half a year now. Makes me wonder why... and how can I upgrade without pulling the git repo...

**Comment by erictraut** [comment]
I think that typing_extensions has been updated to include PEP 612 support, but it sounds like the latest code hasn't been published.

To work around this in the meantime, you might need to resort to hacks like this:

```python
from __future__ import annotations
from typing import Callable, TypeVar, TYPE_CHECKING

if TYPE_CHECKING:
    from typing_extensions import Concatenate, ParamSpec

    _P = ParamSpec("_P")
    _T = TypeVar("_T")
```

**Comment by qumusabel** [comment]
Thank you very much!

**Comment by erictraut** [comment]
I confirmed that the support has been added to `typing_extensions`. Here's [the code](https://github.com/python/typing/blob/40932e3975a1be61d5904f6f790b4631cbaeb74b/typing_extensions/src_py3/typing_extensions.py#L2213) in the `typing_extensions` repo.

@gvanrossum, do you know when a new version of `typing_extensions` will be published to pypi? What is the process for getting this done?

**Comment by gvanrossum** [comment]
> @gvanrossum, do you know when a new version of `typing_extensions` will be published to pypi? What is the process for getting this done?

I'd file an issue requesting one. I've just transferred the responsibility for this job to some new maintainers and they seem eager to get stuff done.
